Generalize volume dampening and allow dampening while not playing

aemino opened the pull request:
I've wanted the feature described in https://github.com/ppy/osu/issues/11298 for a while, so I made an attempt at implementing the proposed changes.

I'm still never sure of the idiomatic way to implement things, so consider this a first draft. Also, of course, feel free to bikeshed the naming of things (including the user-facing text), since I wasn't sure about a lot of it. I changed "dim" to "dampen" as I felt that it was a term better suited for audio, but if this is disagreeable I am more than happy to use the former.

Of potential concern in this implementation are two things:
(1) a slight lack of DRYness, though the "duplication" didn't seem too unreasonable to me
(2) the way that `volumeFade` is updated when the dampened volume is changed means that it is technically possible for it to conflict with an active transition. If this is an issue, please suggest a preferred way to deal with this, since I am unsure myself.

aemino committed:
Generalize volume dampening and allow dampening while not playing

omgisthathim commented:
OMG IS THAT HIM? AEMINO IS SO HOT IRL IS AN ACTUAL CHAD

aemino committed:
Remove redundant this

peppy commented:
There's multiple issues I see:

- It's not implemented in that way I proposed in the linked issue.
- It's dimming interface sounds
- It's applying everywhere. This is the most weird UX I've ever heard of and may not be something we want to apply to a... *rhythm game*
- It's applying in breaks
- It's changing config enum elements, which will destroy existing user configs

80% against this existing in the first place. still agree that it may be something we want in multiplayer though.

aemino commented:
I recognize that it's rather strange that this applies to everything. That said, the reason I find this useful is because I (and others I know) often manually lower music volume while not playing and increase it while playing. This would be a nice way to have that done automatically.

I also agree that it shouldn't apply to things like interface sounds, but given my limited knowledge of the codebase, I didn't want to attempt adding a mechanism for manipulating specific categories of sounds (I think that's sort of what https://github.com/ppy/osu-framework/pull/4291 intends to do?)

Other than this facet, is there something else I missed in your proposal? Regarding the second to last point, I don't quite understand what the implication of the dampening applying in breaks is. Regarding the last point, I can revert the change to the name of `VolumeInactive`; would this avoid breaking configs?

peppy commented:
Second last point: It's a shocking user experience to have the volume of the track getting louder and software multiple times in a play session. There are some beatmaps with rapid break sections which would cause epileptic level volume changes, too.

aemino commented:
I see what you mean wrt. break time. I didn't realize this was an issue because my assumption about the semantics of `LocalUserPlaying` was wrong; I assumed that it merely indicated whether the `Player` was active or not, when in reality it's more nuanced. Would exposing a new `Bindable<bool>` similar to `LocalUserPlaying` but to instead to indicate whether or not a beatmap is currently being played be acceptable, or should I pursue a different solution?

frenzibyte commented:
A flags enumeration could do, to save having another bindable:
```csharp
[Flags]
public enum LocalUserPlayingState
{
    NotPlaying = 0,
    InPlayer = 1,
    TakingBreak = 1 << 1,
}
// confine = InPlayer & ~TakingBreak
```

peppy commented:
I'm still not sure. What about when the user is selecting a song at song select? Is the expectation that they can't hear the song because it's dimmed to 25%? I don't understand.

frenzibyte commented:
I agree. In song select, the volume should not dim at all, as you're selecting a song yourself to be playing it, where you expect to hear it at full volume.

I would say volume dampening/dimming should only happen while inside a multiplayer match, where it makes sense for the beatmap volume to be dimmed a bit while the participants are waiting, until the match starts.

peppy commented:
As an aside:

![20210515 153458 (Safari app)](https://user-images.githubusercontent.com/191335/118350688-20cc1e00-b593-11eb-88a9-61caf39cf61d.png)

![20210515 153519 (Safari app)](https://user-images.githubusercontent.com/191335/118350696-288bc280-b593-11eb-8c18-d0a8e8939b08.png)

If it's decided to only be at multiplayer lobby, it does not need to be a setting. It can probably just be automatic.

aemino commented:
Thinking about this some more, I agree that the approach in this PR is probably more wide-reaching than necessary. When I get a chance next, I can attempt to reimplement this as a multiplayer-only, music-only volume dampen; does that sound agreeable? I think that this addresses most of the concerns raised with the current approach.

Regarding the verbs, I used "dampen" because in my subjective experience, I associate the word with a general reduction of a quality (particularly sound), whereas I associate "dim" almost exclusively with lighting. I don't intend to die on this hill though, and as I noted initially, if you would prefer the usage of "dim" please let me know and I will do so.

smoogipoo commented:
Please continue further discussion in https://github.com/ppy/osu/issues/11298, because we've already determined this isn't the way to go here.
